portality/scripts/article_duplicate_analysis.py

```diff
@@ -465,6 +495,87 @@ def _read_match_set(reader, next_row):
     # a catch to make sure that everything is ok with the match set detection
     assert n_matches + 1 == len(match_set.matches)
 
+
+def finalise(source, report_out, articles_dir, final_actions):
+    if not os.path.exists(articles_dir):
+        os.makedirs(articles_dir)
+
+    actions = ActionRegister()
+    with codecs.open(source, "rb", "utf-8") as s:
+        reader = clcsv.UnicodeReader(s)
+        headers = reader.next()
+
+        accounts = {}
+        for row in reader:
+            article_id = row[0]
+            article_doi = row[2]
+            article_ft = row[3]
+            article_owner = row[4]
+            match_type = row[8]
+            match_id = row[9]
+            match_doi = row[11]
+            match_ft = row[12]
+            match_owner = row[13]
+
+            actions.set_action(article_id, "delete", "could not be automatically cleaned up")
+            actions.set_action(match_id, "delete", "could not be automatically cleaned up")
+
+            if article_owner not in accounts:
+                accounts[article_owner] = []
+            reason = ""
+            if match_type == "doi":
+                reason = "DOI appears in multiple articles"
+            elif match_type == "fulltext":
+                reason = "Fulltext URL appears in multiple articles"
+            else:
+                reason = "Fulltext URL and DOI both appear in multiple articles"
+            accounts[article_owner].append([article_doi, article_ft, reason])
+
+            if match_owner not in accounts:
+                accounts[match_owner] = []
+            reason = ""
+            if match_type == "doi":
+                reason = "DOI appears in multiple articles"
+            elif match_type == "fulltext":
+                reason = "Fulltext URL appears in multiple articles"
+            else:
+                reason = "Fulltext URL and DOI both appear in multiple articles"
+            accounts[article_owner].append([match_doi, match_ft, reason])
+
+    final_instructions = {}
+    actions.export_to(final_instructions)
+
+    with codecs.open(final_actions, "wb", "utf-8") as fa:
+        fawriter = clcsv.UnicodeWriter(fa)
+        fawriter.writerow(["id", "action", "reason"])
+        for k, v in final_instructions.iteritems():
+            fawriter.writerow([k, v["action"], v["reason"]])
+
+    with codecs.open(report_out, "wb", "utf-8") as ro:
+        writer = clcsv.UnicodeWriter(ro)
+        writer.writerow(["account", "articles to delete", "article_details"])
+        for k, v in accounts.iteritems():
+            fn = k + "_articles.csv"
+            with codecs.open(os.path.join(articles_dir, fn), "wb", "utf-8") as a:
+                awriter = clcsv.UnicodeWriter(a)
+                awriter.writerow(["DOI", "Fulltext", "Reason for removal", "Number of duplicated articles"])
+                dedupe = []
+                for article in v:
```

Comment:
I'm finding this loop a little tricky to review. Can you explain it?

Reply:
We're writing two related CSVs at the same time:

1. lists the account name, the number of articles being deleted, and the location of the CSV listing those articles
2. lists the identifers, reason for removal, and number of articles sharing those identifiers.  There is one CSV of this type per account.

These CSVs will be used by the DOAJ team to communicate with the publishers listed in spreadsheet (1) with the details in the related second spreadsheet.

Above we've gathered the data on the accounts and the articles that will be deleted.  This is in a datastructure like:

```
accounts = {
    "account_name" : [
        ["doi", "fulltext", "reason"],
        ...
    ]
}
```

All this loop is doing is flattening that structure so that a number of rows can be written to the second sheet, and a single row can be written to the first sheet.  The only additional complexity is that since ids may be repeated in each set of id records, they must first be deduplicated.  So by the end of the loop, you have a variable called `dedupe` which contains:

```
[
    ["unique doi 1", "unique fulltext 1", "reason", "count of records with these identifiers"],
    ["unique doi 2", "unique fulltext 2", "reason", "count of records with these identifiers"]
]
```

These rows are then written to the second spreadsheet.

Finally, the `account_name` referred to as `k`, the number of records `len(v)`, and the filename `fn` of the second spreadsheet are written to the first.  We use the number of records `len(v)` rather than `len(dedupe)` because the former tells us how many actual articles there are.

Does that help?

Reply:
fantastic, thank you!